indico/web/client/js/react/components/DateRangePicker.jsx

```diff
@@ -0,0 +1,224 @@
+// This file is part of Indico.
+// Copyright (C) 2002 - 2024 CERN
+//
+// Indico is free software; you can redistribute it and/or
+// modify it under the terms of the MIT License; see the
+// LICENSE file for more details.
+
+import moment from 'moment';
+import PropTypes from 'prop-types';
+import React from 'react';
+
+import DatePickerCalendar from 'indico/react/components/DatePickerCalendar';
+import {FinalField, validators as v} from 'indico/react/forms';
+import {Param, Translate} from 'indico/react/i18n';
+import {DateRange} from 'indico/utils/date';
+import {formatDate} from 'indico/utils/date_format';
+import {fromISOLocalDate} from 'indico/utils/date_parser';
+
+import 'indico/custom_elements/ind_date_picker';
+
+const INVALID = '__invalid__';
+
+export default function DateRangePicker({
+  value,
+  label,
+  format,
+  rangeStartLabel,
+  rangeEndLabel,
+  rangeStartMin,
+  rangeStartMax,
+  rangeEndMin,
+  rangeEndMax,
```

Comment:
I think this is very nice for simple cases where you have a fixed start and end date.

However, especially in Room Booking, we have some more complex cases where a callback to determine whether a given date is valid or not is much more convenient:

For example, we have this in `BookingBootstrapForm`:

```js
isOutsideRange={dt => !isBookingStartDateValid(dt, isAdminOverrideEnabled, bookingGracePeriod)}
```

and the logic behind it is this:

```js
export function isBookingStartDateValid(date, isAdminOverrideEnabled, gracePeriod) {
  if (!date || !date.isValid()) {
    return false;
  } else if (isAdminOverrideEnabled) {
    return true;
  } else if (gracePeriod === null) {
    return date.isSameOrAfter(moment(), 'day');
  }
  return date.isSameOrAfter(moment().subtract(gracePeriod, 'hour'), 'day');
}
```

I tried this which does mostly the same:

```js
    const minDate = isAdminOverrideEnabled
      ? ''
      : bookingGracePeriod === null
      ? '2024-08-22' // that's today, at least today ;)
      : '2024-08-22'; // too lazy to do the hour-based calculation here for testing
```

But for example, when I toggle admin override (ctrl-shift-a) after the picker has been initialized, it will not take the new value into account at all. Likewise when the picker is already open and I use that shortcut to toggle it, it's not taken into account (while this worked fine w/ the old one).


FWIW here's the full patch I'm using locally to test this in the room booking module:

<details>

```patch
diff --git a/indico/modules/rb/client/js/components/BookingBootstrapForm.jsx b/indico/modules/rb/client/js/components/BookingBootstrapForm.jsx
index aee23910b3..d6adb591c3 100644
--- a/indico/modules/rb/client/js/components/BookingBootstrapForm.jsx
+++ b/indico/modules/rb/client/js/components/BookingBootstrapForm.jsx
@@ -13,7 +13,7 @@ import Overridable from 'react-overridable';
 import {connect} from 'react-redux';
 import {Button, Form, Select} from 'semantic-ui-react';

-import {SingleDatePicker, DateRangePicker} from 'indico/react/components';
+import {SingleDatePicker, DateRangePicker2} from 'indico/react/components';
 import {PluralTranslate, Translate} from 'indico/react/i18n';
 import {
   serializeDate,
@@ -320,6 +320,14 @@ class BookingBootstrapForm extends React.Component {
     );
     const recurrenceHidden = hideOptions.recurringWeekly && hideOptions.recurringMonthly;

+    const minDate = isAdminOverrideEnabled
+      ? ''
+      : bookingGracePeriod === null
+      ? '2024-08-22'
+      : '2024-08-22'; // XXX too lazy to do the hour-based calculation here
+
+    console.log(minDate);
+
     return (
       <Form>
         {showRecurrenceOptions && (
@@ -379,10 +387,14 @@ class BookingBootstrapForm extends React.Component {
         )}
         {['every', 'daily'].includes(type) && (
           <Form.Group inline>
-            <DateRangePicker
+            <DateRangePicker2
+              rangeStartLabel="Start"
+              rangeEndLabel="End"
               startDate={startDate}
               endDate={endDate}
-              onDatesChange={({startDate: sd, endDate: ed}) => this.updateDates(sd, ed)}
+              rangeStartMin={minDate}
+              rangeEndMin={minDate}
+              onChange={({startDate: sd, endDate: ed}) => this.updateDates(sd, ed)}
               isOutsideRange={dt => {
                 return !isBookingStartDateValid(dt, isAdminOverrideEnabled, bookingGracePeriod);
               }}
```

</details>

Reply:
So if I understand correctly, it would be sufficient to make the new date range picker react to changes in the rangeStart and rangeMin? If so, what would be the desired behavior in case the currently selected rangeStart is not within that range? Deselect it?

Reply:
Ok, scratch this comment. I saw your later comment below.

Reply:
So here's what it looks like to me:

# `WTFDateField`

Current implementation uses the isOutsideRange prop in react-date:

```
  const isOutsideRange = useCallback(
    value => {
      if (linkedField) {
        const linkedDate = moment(linkedFieldDateElem.value, 'DD/MM/YYYY');
        if (
          (linkedField.notBefore && value.isBefore(linkedDate, 'day')) ||
          (linkedField.notAfter && value.isAfter(linkedDate, 'day'))
        ) {
          return true;
        }
      }
      return (
        (earliestMoment && value.isBefore(earliestMoment, 'day')) ||
        (latestMoment && value.isAfter(latestMoment, 'day'))
      );
    },
    [earliestMoment, latestMoment, linkedField, linkedFieldDateElem]
  );
```


This translates to:

```
linkedField =>
  linkedField.notBefore => min = linkedField.value
  linkedField.notAfter => max = linkedField.value

not linkedField => 
  earliestMoment => min = earliestMoment
  latestMoment => max = latestMoment
```

Note that `DatePicker` (not range) only has `min` and `max`.

# `BookingBootstrapForm`

Current implementation uses:

```
  dt => !isBookingStartDateValid(dt, isAdminOverrideEnabled, bookingGracePeriod)

  function isBookingStartDateValid(date, isAdminOverrideEnabled, gracePeriod) {
    if (!date || !date.isValid()) {
      return false;
    } else if (isAdminOverrideEnabled) {
      return true;
    } else if (gracePeriod === null) {
      return date.isSameOrAfter(moment(), 'day');
    }
    return date.isSameOrAfter(moment().subtract(gracePeriod, 'hour'), 'day');
  }
```

If I'm understanding this correctly, it only limits the start date, so effectively behaving as `min` in `DatePicker`, or, in other words, `rangeStartMin = rangeStartMax` in `DateRangePicker`.

This translates roughly to:

```
isAdminOverrideEnabled => rangeStartMin = rangeStartMax = ''

not isAdminOverrideEnabled =>
  gracePeriod => rangeStartMin = rangeStartMax = getDateOf(now - gracePeriod)
  not gracePeriod => rangeStartMin = rangeStartMax = now
```

# Conclusion

1. There's technically no need for a callback.
2. There's an obvious need to simplify the interface for cases where `range*Min = range*Max` at the expense of some additional complexity in the underlying custom element. Using the accommodation field use case as the base case was not a good idea.
3. There's a need to make `range*Min` and `range*Max` dynamic, so that the calendars will reflect a change in those values.

A callback might be useful for expressing things like "Is this date available?" but not for limiting ranges, imo. It obscures the intent of the code. For instance, with callbacks, we are forced to express the concepts of `min` and `max` in terms of date comparison, which is harder to read than just saying `min = X` or `{min: X}`. In both examples of existing callbacks, I was not able to quickly tell which side of the range is being limited by reading the code. At this stage, I don't see a case nor a need for a callback.

To simplify the interface we could add `min` and `max` props that will be used as a fallback when `rangeStart*` and `rangeEnd*`, respectively, are not specified. The developer can then pick the level of detail they would like to express without resorting to double-assignment.

Reply:
But the linked field would not trigger a re-render of the react component, so the min/max wouldn't update when that field changes...

Reply:
But *maybe* it's enough if you make sure that a change to the min/max-related props causes the picker to update immediately...

Reply:
> But the linked field would not trigger a re-render

I think this can be solved by tracking the linked field value as state. Something like this:

```
const [linkedFieldValue, setLinkedFieldValue] = useState()

useEffect(() => {
  const abortCtrl = new AbortController()

  document.getElementById(`${linkedField.id}-datestorage`)
    .addEventListener('input', evt => {
      setLinkedFieldValue(evt.target.value)
    }, {signal: abortCtrl.signal})

  return () => abortCtrl.abort()
}, [linkedField])
```

Reply:
I've implemented the changes to make the both data pickers react to range changes.

I've also added aliases `min` and `max` to the `DateRangePicker` component, such that they can be used instead of assigning identical values to `rangeStart*` and `rangeEnd*`.

I didn't implement the hack to make WTFDateField react to the linkedField, tho.

Anyhow, let me know if this works for you.

Reply:
Not sure how much sense those min/max shorthands make...

Using them will, in the end, restrict the start/end to an exact valueshorthands currently allow selecting *only* that particular date, because it's both the min and max of the range.

Shouldn't it work like this?

- set `rangeStartMin=min` and `rangeStartMax=''` for `min` so the lowest (start) date one can select is `min`
- set `rangeEndMin=''` and `rangeEndMax=max` for `max` so the highest (end) date one can select is `max`


Anyway, right now they don't appear to be super useful so maybe better to get rid of them? I don't see much benefit in having them (neither in general, nor in context of the discussions above) and it may be easier to add them back later if we come up with a real usecase for them, instead of having something weird where people will later wonder why it exists..

Reply:
BTW you were right that it's much cleaner to avoid using callbacks for the earliest possible date in room booking.

Reply:
> Anyway, right now they don't appear to be super useful so maybe better to get rid of them?

Ok. Let's get rid of it.

> Shouldn't it work like this?

I'll think about it tomorrow. It sounds very logical at first glance, but I might need to deal with some edge cases or adjust the underlying Range classes.

Reply:
I think when getting rid of it there's no need to think about it for now ;)